README.md
# RuneLiteBot
Disclaimer: I do not condone cheating. I simply wanted to improve my Python skills, 
and apply my software development knowledge to a game I used to love growing up as a kid.


## Description
Programmed in Python and used in many desktop applications, I decided to create a bot script for the RuneLite client in Old School Runescape without the user having to control the mouse and keyboard to level up the user's woodcutting skill. More functionality to be added soon.


## Installation Requirements
Only supported for Windows OS. First Google and download Python 3.9.5.

Once installed, go to your favorite terminal (Command Prompt):
Python 3.9.5 - Python3 by default should have `pip` already installed.
`pip install pyautogui`
`pip install pygetwindow`
`pip install Pillow`

The following libraries install PyAutoGui, PyGetWindow, and Pillow (Image Capture) respectively.


## How it Works
This is currently a woodcutting bot - perfect for someone whose level is between 1-30 for cutting regular and oak trees.

1. Login to RuneLite (Make sure your game window is not fullscreen)
2. Your Game client layout should be: `Resizable - Classic Layout`
3. Your top 4 slots in your inventory should be free. Weapons should be at bottom of inventory.

Make sure your brightness level is set at 2nd knotch.
The following screenshot will display the optimal settings.

Please refer to the following screenshots below for Settings and Inventory.

![Display Settings](/images/DisplaySettings.png) ![Inventory](/images/Inventory.png)

Note: Don't expand the RuneLite sidebar. 

Once Python and its required libraries are installed, start your favorite console and type: `py main.py`

Changing the Settings for running the script is easy. Refer to the `dev.ini` file.


## Development Settings
Please refer to the following file `dev.ini`
drop_number_of_logs_per_cycle value should between 1 - 4 to follow algorithm cycle


## Tested Settings
- Windows 10, Python 3.9.5, 
- Screen Resolution: 1920 x 1080
Only supported with 100% scaling for Windows.


## Known Bugs:
- [ ] When screen size is classic and too small, program doesn't execute likely due to failure of image capture attempting to find a random tree algorithm


## Resources
https://realpython.com/python-gui-tkinter/
https://docs.python.org/3/library/configparser.html

changes.md
# Version Changes:

## Version 1.0 Features:
- [x] Improve findRandomTree() algorithm where bot may accidently mistake a log for a tree
    - Implemented confirmTree() algorithm
    - Pixel color detection for finding tree outside of inventory - this approach speeds up algorithm slightly 
- [x] Create Python script for additional support
- [x] Create time variable if user isn't able to detect tree within max number of attempts

## Version 1.1 Features:
- [x] Implement basic configuration file
    - Prevents end-user from having to key in settings repeatedly
- [x] Create first rough draft of the OSRSBot UI
    - Reads existing configuration file
    - Implement both START button 
    - Doesn't fully function due to lack of asynchronous programming
- [x] Add supporting images and videos for better understanding of documentation
- [x] Refactored code for Main execution script for OOP principles
- [x] Implement basic configuration support for number of logs to be dropped per cycle

## TODO Additional Features:
- [ ] Include Youtube demo video
- [ ] Create another findTree() algorithm where color detection starts from center of window??
- [ ] OSRS main client support?
- [ ] Support Window Scaling

## TODO Features for Version 2:
- [ ] Woodcutting Bot
    - [ ] Implement OpenCV
    - [ ] Utilize TensorFlow machine learning algorithm
    - [ ] Change woodcutting bot cycle
- [ ] Other bots - mining or firemaking
- [ ] Write basic system and unit tests using PyTest
- [ ] Improve Usability
    - [ ] Enhance UI
    - [ ] Create an executable Windows exe file?
    - [ ] Integrate GUI together with concurrency support using asynchronous programming


main.py
# Installed default dependencies
from configparser import ConfigParser   # Handles configuration file settings
from time import sleep                  # Async-await

# Required dependencies
import pyautogui                        # Automated tools - Pixel matching not supported in this version
import pygetwindow as gw                # Simplified version for win32gui library

# Helpers
from helpers import woodcutting_bot_helper

# Class responsible for executing program logic
class Main:

    def setup(self):
        parser = ConfigParser()
        parser.read('dev.ini')

        selectionOne = parser.get('settings','bot_type',fallback='none')
        selectionTwo = parser.getint('settings','woodcutting_find_tree_attempts',fallback=20)
        selectionThree = parser.getint('settings','chop_and_drop_time_seconds',fallback=15)
        selectionFour = parser.getint('settings', 'drop_number_of_logs_per_cycle',fallback=4) 
                
        # TODO: Validate data
        if (selectionOne == 'woodcutting'):
            if (selectionTwo >= 1 and selectionTwo <= 1000):
                self.onInit(selectionTwo,selectionThree,selectionFour)
            else:
                print('Please make sure woodcutting_find_tree_attempts is between 1 to 1000')
                exit()
        elif (selectionOne == 'surprise'):
            self.onInitTest()
        else:
            print('Please check your configuration file again!')
            exit()

    def onInit(self,attempts,timeSeconds,numberOfLogs):

        print('The program is starting!\nPress Control+C in the console to exit the program anytime.')
        sleep(2.000)

        numberConsecutiveAttempts = 0
        maxConsecutiveNumberAttempts = int(attempts)

        while (numberConsecutiveAttempts < maxConsecutiveNumberAttempts):

            print('Player finding tree!')
            sleep(0.250)
            tree = woodcutting_bot_helper.findRandomTreeAlgorithm()

            if (not tree):
                woodcutting_bot_helper.rotateCamera()
                numberConsecutiveAttempts = numberConsecutiveAttempts + 1
                continue

            print('The tree is located at ' + str(tree['x']) + ',' + str(tree['y']))
            pyautogui.moveTo(tree['x'], tree['y'])

            numberConsecutiveAttempts = 0           # Resets to 0

            sleep(0.300) # Remove possible misclicks
            pyautogui.leftClick()

            sleep(timeSeconds) # Additional time waited
            woodcutting_bot_helper.dropLogs(numberOfLogs)

    def onInitTest(self):
        print('You got trolled lmao')

# Main Class Execution
main = Main()
main.setup()

helpers/woodcutting_bot_helper.py
# Required dependencies
from time import sleep              # Async-await
import pyautogui                    # Automated tools - Pixel matching not supported in this version
import pygetwindow as gw            # Simplified version for win32gui library
from PIL import ImageGrab    # Image dev tools
import random
from helpers import helper

def confirmTree(screen_x,screen_y):

    OSRSWindow = gw.getWindowsWithTitle('RuneLite')[0]

    x = OSRSWindow.topleft[0]
    y = OSRSWindow.topleft[1]
    width = OSRSWindow.width
    height = OSRSWindow.height # Values to be passed as arguments; used for debugging only

    pyautogui.moveTo(screen_x,screen_y)
    img = ImageGrab.grab(bbox=(x,y,width,height))

    cyan_color = (0,200,200)

    # Coordinates may slightly differ
    for y_coord in range(41,43,1):
        for x_coord in range(105,112,1):
            pixel_color = img.getpixel((x_coord,y_coord))
            print('At x ' + str(x_coord) + ' and at y ' + str(y_coord) + ' color: ' + str(pixel_color))
            if (helper.is_similar(pixel_color, cyan_color, 20)):
                return True

    return False

def findRandomTreeAlgorithm():

    OSRSWindow = gw.getWindowsWithTitle('RuneLite')[0]

    x = OSRSWindow.topleft[0]
    y = OSRSWindow.topleft[1]
    width = OSRSWindow.width
    height = OSRSWindow.height # Values to be passed as arguments; used for debugging only

    img = ImageGrab.grab(bbox=(x,y,width,height)) # Doesn't save image produced

    # Sample list of color tuples with form of RGB 
    tree_trunk_colors = [(74,50,23),(91,61,28),(94,64,29),(93,63,29)]

    inventory_size = {'x': 320, 'y': 422} # Inventory size + sidebar calculation not accurate

    for i in range(0,3000,1):
        random_x = random.randrange(0, img.width - inventory_size['x'] - 1) # Ignore detection of inventory logs
        random_y = random.randrange(0, img.height - inventory_size['y'] - 1) # This ultimately reduces search space
        #print(str(random_x) + ',' + str(random_y))

        sample_color = img.getpixel((random_x,random_y))

        if (sample_color in tree_trunk_colors):
            screen_x = random_x + x
            screen_y = random_y + y

            img = ImageGrab.grab(bbox=(x,y,width,height))
            if (confirmTree(screen_x,screen_y)):
                print('Confirmed tree at following coords: ' + str(screen_x) + ',' + str(screen_y) + ' and its color value is ' + str(sample_color))
                print('The tree has not been cut down yet! Cutting down the tree...')
                return { "x": screen_x, "y": screen_y }
            else:
                print('Unconfirmed tree at following coords: ' + str(screen_x) + ',' + str(screen_y) + ' color ' + str(sample_color))

    return False

def dropLogs(numberOfLogs):

    OSRSWindow = gw.getWindowsWithTitle('RuneLite')[0]

    inventory_x = OSRSWindow.bottomright[0] - 228 # Test for scaling
    inventory_y = OSRSWindow.bottomright[1] - 280 # Test for scaling

    for times_x in range(1, numberOfLogs+1):

        pyautogui.moveTo(inventory_x,inventory_y)
        pyautogui.rightClick()
        pyautogui.moveTo(inventory_x,inventory_y + 40)
        pyautogui.leftClick()

        if (times_x < 4):
            inventory_x = inventory_x + 40 # Test for scaling
            sleep(0.5)

        times_x = times_x + 1

# This function can technically be classified in another class for future development
def rotateCamera():
    pyautogui.keyDown("right")
    sleep(0.500)
    pyautogui.keyUp("right")


testMain.py
## This is the executable file with (somewhat poor) UI Support

# Installed default dependencies
from functools import partial
import tkinter as tk
from tkinter import IntVar, StringVar, Text, ttk                 # Used for creating GUI application
from configparser import ConfigParser   # Handles configuration file settings
from time import sleep                  # Async-await

# Required dependencies
import pyautogui                        # Automated tools - Pixel matching not supported in this version
import pygetwindow as gw                # Simplified version for win32gui library

# Helpers
from helpers import woodcutting_bot_helper

def setupApp():
    # Load and Read Existing Settings
    parser = ConfigParser()
    parser.read('dev.ini')

    selectionOne = parser.get('settings','bot_type',fallback='none')
    selectionTwo = parser.getint('settings','woodcutting_find_tree_attempts',fallback=0)
    selectionThree = parser.getint('settings','chop_and_drop_time_seconds',fallback=15)

    # Create RuneLiteBot Application Window
    window = tk.Tk()
    window.geometry('480x320')

    # Create Labels
    ttk.Label(window, text="OSRS RuneLite Bot",
    font = ("Times New Roman", 12)).grid(column=0,row=0)
    ttk.Label(window, text="Select the Bot Type: ",
        font = ("Times New Roman", 12)).grid(column=0,row=1)
    ttk.Label(window, text="Woodcutting: Find Tree Attempts",
        font = ("Times New Roman", 12)).grid(column=0,row=3)
    ttk.Label(window, text="Woodcutting: Chop and Drop Time",
        font = ("Times New Roman", 12)).grid(column=0,row=4)

    v1 = IntVar()
    e1 = ttk.Entry(window, textvariable=v1)
    v1.set(selectionTwo)
    e1.grid(column=1,row=3)

    v2 = IntVar()
    e2 = ttk.Entry(window, textvariable=v2)
    v2.set(selectionThree)
    e2.grid(column=1,row=4)

    # Combobox Creation
    n = tk.StringVar()
    bot_type_selection = ttk.Combobox(window, width=27, textvariable=n)
    # Add combobox dropdown list
    bot_type_selection['values'] = ('woodcutting','surprise','none')
    bot_type_selection.grid(column=1,row=1)
    bot_type_selection.current(0)
    
    # Basic START Button (Needs better support)
    startBtn = ttk.Button(window, text="START", command=partial(onInit,v1.get(),v2.get())).grid(column=0,row=5)
    stopBtn = ttk.Button(window, text="STOP", command=exit).grid(column=0,row=6)

    # Keep window going
    window.mainloop()


def setup():
    parser = ConfigParser()
    parser.read('dev.ini')

    selectionOne = parser.get('settings','bot_type',fallback='none')
    selectionTwo = parser.getint('settings','woodcutting_find_tree_attempts',fallback=20)
    selectionThree = parser.getint('settings','chop_and_drop_time_seconds',fallback=15)
    if (selectionOne == 'woodcutting'):
        if (selectionTwo >= 1 and selectionTwo <= 1000):
            onInit(selectionTwo,selectionThree)
        else:
            print('Please make sure woodcutting_find_tree_attempts is between 1 to 1000')
            exit()
    elif (selectionOne == 'surprise'):
        onInitTest()
    else:
        print('Please check your configuration file again!')
        exit()

def onInit(attempts,timeSeconds):

    print('The program is starting!\nPress Control+C in the console to exit the program anytime.')
    sleep(2.000)

    numberConsecutiveAttempts = 0
    maxConsecutiveNumberAttempts = int(attempts)

    while (numberConsecutiveAttempts < maxConsecutiveNumberAttempts):

        print('Player finding tree!')
        sleep(0.250)
        tree = woodcutting_bot_helper.findRandomTreeAlgorithm()

        if (not tree):
            woodcutting_bot_helper.rotateCamera()
            numberConsecutiveAttempts = numberConsecutiveAttempts + 1
            continue

        print('The tree is located at ' + str(tree['x']) + ',' + str(tree['y']))
        pyautogui.moveTo(tree['x'], tree['y'])

        numberConsecutiveAttempts = 0           # Resets to 0

        sleep(0.300) # Remove possible misclicks
        pyautogui.leftClick()

        sleep(timeSeconds) # Additional time waited
        woodcutting_bot_helper.dropLogs()

def onInitTest():
    print('You got trolled lmao')

# Application running in multiple threads doesn't quite work yet
setupApp()
# setup()